gustaf/io/npz.py

```diff
@@ -0,0 +1,170 @@
+    type must match what is found in 'kind'.
+
+    Arrays, such as 'vertices', are read from the file as they are. Dictionaries
+    are created from certain array groups that start with the same prefix, such
+    as 'vertex_groups_'.
+
+    Parameters
+    -----------
+    fname : str
+    force_type : type or None
+        (Default: None)
+    check : bool
+        (Default: False) Check consistency of connectivity data.
+
+    Returns
+    -------
+    mesh : Volumes, Faces, Edges, or Vertices
+    """
+    # read a dictionary from the file
+    with np.load(fname, allow_pickle=False) as data:
+        # find kind and set mesh class type
+        mesh_kind_array = data.get("kind", None)
+        if mesh_kind_array is not None and len(mesh_kind_array) == 1:
+            mesh_kind = mesh_kind_array[0]
+        else:
+            raise RuntimeError("NPZ file is missing correct 'kind' array. "
+                    "Found " + ", ".join(data) + ".")
+
+        # prepare kind => type dict (e.g. 'volume': Volumes)
+        kind_to_type = {type_.kind: type_ for type_ in _stored_arrays.keys()}
+        try:
+            mesh_type = kind_to_type[mesh_kind]
+        except KeyError:
+            raise RuntimeError(f"Kind '{mesh_kind}' doesn't match any known "
+                    + "types: " + ", ".join(kind_to_type) + ".")
+
+        if force_type is not None and force_type != mesh_type:
+            raise RuntimeError(f"Requested {force_type} but got {mesh_type}.")
+
+        # loop over all array entries and assign correctly
+        # (we're importing into intermediate variables so we can make sure all
+        # arrays are imported before the dicts.)
+        read_arrays = {}
+        read_dicts = {}
+        for name, content in data.items():
+            # check for array
+            if name in _stored_arrays[mesh_type]:
+                read_arrays[name] = content
+            elif name == "kind":
+                continue
+            else:
+                # read dictionary entry name as list to avoid error if
+                # underscore is missing
+                dict_name, *dict_entry_name = name.split("_", 1)
+                # check for dict
+                if (dict_name in _stored_dicts[mesh_type]
+                        and len(dict_entry_name) > 0):
+                    if dict_name not in read_dicts:
+                        read_dicts[dict_name] = {}
+                    read_dicts[dict_name][dict_entry_name[0]] = content
+                else:
+                    utils.log.warning("Encountered unknown array "
+                            f"'{name}' while reading NPZ file.")
+
+        # check if everything's there
+        missing_arrays = set(_stored_arrays[mesh_type]).difference(
+                set(read_arrays))
+        if missing_arrays:
+            raise RuntimeError("Missing arrays in NPZ file: "
+                    + ", ".join(list(missing_arrays)))
```

Comment:
can we replace this with warning? for example, faces without connectivity can be faces if they want